core/primitives/src/time.rs

```diff
@@ -1,155 +1,295 @@
-        F: FnOnce(&mut MockClockPerThread) -> T,
-    {
-        thread_local! {
-            static INSTANCE: RefCell<MockClockPerThread> = RefCell::default()
+    pub struct Clock {}
+
+    impl MockClockPerThread {
+        pub fn reset(&mut self) {
+            self.utc.clear();
+            self.durations.clear();
+            self.utc_call_count = 0;
+            self.instant_call_count = 0;
+            self.instant = Instant::now();
+            self.is_mock = false;
+        }
+
+        fn with<F, T>(f: F) -> T
+        where
+            F: FnOnce(&mut MockClockPerThread) -> T,
+        {
+            thread_local! {
+                static INSTANCE: RefCell<MockClockPerThread> = RefCell::default()
+            }
+            INSTANCE.with(|it| f(&mut *it.borrow_mut()))
+        }
+
+        fn pop_utc(&mut self) -> Option<DateTime<chrono::Utc>> {
+            self.utc_call_count += 1;
+            self.utc.pop_front()
+        }
+        fn pop_instant(&mut self) -> Option<Instant> {
+            self.instant_call_count += 1;
+            let x = self.durations.pop_front();
+            match x {
+                Some(t) => self.instant.checked_add(t),
+                None => None,
+            }
         }
-        INSTANCE.with(|it| f(&mut *it.borrow_mut()))
     }
 
-    fn pop_utc(&mut self) -> Option<DateTime<chrono::Utc>> {
-        self.utc_call_count += 1;
-        self.utc.pop_front()
+    impl Default for MockClockPerThread {
+        fn default() -> Self {
+            Self {
+                utc: VecDeque::with_capacity(16),
+                durations: VecDeque::with_capacity(16),
+                utc_call_count: 0,
+                instant_call_count: 0,
+                instant: Instant::now(),
+                is_mock: false,
+            }
+        }
     }
-    fn pop_instant(&mut self) -> Option<Instant> {
-        self.instant_call_count += 1;
-        let x = self.durations.pop_front();
-        match x {
-            Some(t) => self.instant.checked_add(t),
-            None => None,
+
+    pub struct MockClockGuard {}
+
+    impl Default for MockClockGuard {
+        fn default() -> Self {
+            Clock::set_mock();
+            Self {}
         }
     }
-}
 
-impl Default for MockClockPerThread {
-    fn default() -> Self {
-        Self {
-            utc: VecDeque::with_capacity(16),
-            durations: VecDeque::with_capacity(16),
-            utc_call_count: 0,
-            instant_call_count: 0,
-            instant: Instant::now(),
-            is_mock: false,
+    impl Drop for MockClockGuard {
+        fn drop(&mut self) {
+            Clock::reset();
         }
     }
-}
 
-pub struct MockClockGuard {}
+    impl Clock {
+        pub fn set_mock() {
+            MockClockPerThread::with(|clock| {
+                clock.is_mock = true;
+            });
+        }
+        pub fn reset() {
+            MockClockPerThread::with(|clock| {
+                clock.reset();
+            });
+        }
```

Comment:
Seems like these two methonds should be private, as one always must use the guard to properly restore the clock.

Reply:
https://github.com/near/nearcore/pull/5362